File: archived/v1/src/go/ch01/1.2/l1/modulo_examples.go
package main

import "fmt"

// hashingExample maps a key into a table of given size.
func hashingExample(key, size int) int { return key % size }

// cyclicDay wraps weekday (0–6) with a shift.
func cyclicDay(start, shift int) int { return (start + shift) % 7 }

// modexp computes (a^b) mod m using square-and-multiply.
func modexp(a, b, m int) int {
	res := 1
	base := a % m
	exp := b
	for exp > 0 {
		if exp&1 == 1 {
			res = (res * base) % m
		}
		base = (base * base) % m
		exp >>= 1
	}
	return res
}

func runModuloExamples() {
	fmt.Println("Hashing: key=1234, size=10 ->", hashingExample(1234, 10))
	fmt.Println("Cyclic: Saturday(5)+4 ->", cyclicDay(5, 4)) // 2 = Wed
	fmt.Println("modexp(7,128,13) =", modexp(7, 128, 13))   // 3
}

File: archived/v1/src/python/ch01/1.2/l1/modulo_examples.py
def hashing_example(key: int, size: int) -> int:
    return key % size

def cyclic_day(start: int, shift: int) -> int:
    return (start + shift) % 7  # wrap around the week (0–6)

def modexp(a: int, b: int, m: int) -> int:
    """Square-and-multiply modular exponentiation."""
    result = 1
    base = a % m
    exp = b
    while exp > 0:
        if exp & 1:
            result = (result * base) % m
        base = (base * base) % m
        exp >>= 1
    return result

def main():
    print("Hashing: key=1234, size=10 ->", hashing_example(1234, 10))
    print("Cyclic: Saturday(5)+4 ->", cyclic_day(5, 4))  # 5=Saturday
    print("modexp(7, 128, 13) =", modexp(7, 128, 13))


if __name__ == "__main__":
    main()
